data-pipeline/telemetry-location-updater/src/main/java/org/ekstep/ep/samza/util/UserLocationCache.java

```diff
@@ -0,0 +1,112 @@
+        this.cassandra_user_table = config.get("middleware.cassandra.user_table", "user");
+        this.cassandra_location_table = config.get("middleware.cassandra.location_table", "location");
+        this.cassandraConnection = cassandraConnect;
+        this.locationDbKeyExpiryTimeInSeconds = config.getInt("location.db.redis.key.expiry.seconds", 86400);
+    }
+
+    public Location getLocationByUser(String userId) {
+        if (userId == null) return null;
+
+        try (Jedis jedis = redisConnect.getConnection()) {
+            Map<String, String> locationMap = jedis.hgetAll(userId);
+            if (locationMap.isEmpty()) {
+                return fetchUserLocation(userId);
+            } else {
+                return new Location(null, null, null, locationMap.get("state"), null, locationMap.get("district"));
+            }
+        } catch (JedisException ex) {
+            LOGGER.error("", "getLocationByUser: Unable to get a resource from the redis connection pool ", ex);
+            return null;
+        }
+    }
+
+    public Location fetchUserLocation(String userId) {
+        List<Row> rows;
+        Location location = new Location();
+        List<String> locationIds = null;
+
+        if (userId == null) return null;
+        try {
+            String selectLocationIds = QueryBuilder.select("locationids")
+                    .from(cassandra_db, cassandra_user_table)
+                    .where(QueryBuilder.eq("id", userId))
+                    .getQueryString();
+            rows = cassandraConnection.execute(selectLocationIds);
+            if (rows.size() > 0) {
+                Row row = rows.get(0);
+                locationIds = row.getList("locationids", String.class);
+            }
+
+            if (locationIds != null && !locationIds.isEmpty()) {
+                String selectLocations = QueryBuilder.select().all()
+                        .from(cassandra_db, cassandra_location_table)
+                        .where(QueryBuilder.in("id", locationIds)).getQueryString();
+                rows = cassandraConnection.execute(selectLocations);
+                if (rows.size() > 0) {
+                    rows.forEach(record -> {
+                        String name = record.getString("name");
+                        String type = record.getString("type");
+                        if (type.toLowerCase().equals("state")) {
+                            location.setState(name);
+                        } else if (type.toLowerCase().equals("district")) {
+                            location.setDistrict(name);
+                        }
+                    });
+                }
+                addToCache(userId, location);
+                return location;
+            } else {
+                return null;
+            }
+        } catch (Exception ex) {
+            LOGGER.error("", "fetchUserLocation: Unable to fetch user location from cassandra!", ex);
```

Comment:
Add the userId in the exception so that it will be easier to debug.

Reply:
handled.